### ISSUE TITLE
(question) Best practice to wrap new for out of memory exception

### ISSUE THREAD

**Issue by dongahn** [flux-framework/flux-sched#377]
The graph-based resource matching service will be C++ based and thus it will use `new` instead of `malloc`. We wrap `malloc` with `xzmalloc` to capture OOM. What would be a good practice to wrap new?

**Comment by garlick** [comment]
I would say `xzmalloc()` is only appropriate in prototype comms modules since it will tear down or hang the entire instance in the unlikely event that it fails.

If handling a user request, it would be better to return ENOMEM in the response and continue on.

If implementing a library that will be used in multiple contexts, return the error (or propagate the exception) to the caller and let them decide what to do.

**Comment by dongahn** [comment]
But c++ class cannot use xzmalloc. I has to use `new`. I got some ideas.

**Comment by garlick** [comment]
My point was it should not try to do what `xzmalloc()` does (exit) if it's to be used in comms module context.   That is not a best practice that should be carried forward.

**Comment by SteVwonder** [comment]
While testing #464, I built flux with spack (using gcc 6.4.0) on my local desktop.  I'm not sure exactly what went wrong during the build/tests, but during `make check` it resulted in the following error: 
```
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
flux-start: 0 (pid 28332) Aborted 
```

AFAICT, that error shutdown the entire broker.  I believe many of the `new` statements have a `nothrow` associated with them, but I think we may have missed some.  That, or `nothrow` isn't sufficient.  Either way, at first glance, it looks like the `resource` module can still pretty easily teardown the whole broker.  I wonder if there is some way to wrap the entire `mod_main` or `flux_reactor_start` in the `resource` module to catch these errors and "gracefully" handle them (e.g., the resource module unloads itself without killing the broker).

**Comment by dongahn** [comment]
Ugh. Yeah, we should search all new statements and change it to nothrow if we missed some. In fact I fixed one such case in PR #467 yesterday. 

The other possibility is this arises from a dependent library such as yaml-cpp.

>  I wonder if there is some way to wrap the entire mod_main or flux_reactor_start in the resource module to catch these errors and "gracefully" handle them (e.g., the resource module unloads itself without killing the broker).

Given this exception can emerge from a library that we don't have control over. It does make sense to create a catch-all at the top module level.
